builder/docker.go

```diff
@@ -0,0 +1,264 @@
+// in the list. This function requires paths to have no trailing slash.
+func reduceDockerMounts(paths []string) []string {
+	if len(paths) <= 1 {
+		return paths
+	}
+
+	sort.Strings(paths) // Puts "/foo" before "/foo/bar"
+
+	for i := 1; i < len(paths); i++ {
+		if paths[i] == paths[i-1] || strings.HasPrefix(paths[i], paths[i-1]+"/") { // "/" is to prevent "/foobar" matching "/foo"
+			paths = append(paths[:i], paths[i+1:]...)
+			i-- // Because removal shifts things left
+		}
+	}
+
+	return paths
+}
+
+// getDockerMounts returns a minimal list of all directories in the config that
+// need to be mounted inside the container. Only the "Buiilder" and "Mixer"
+// sections of the conf are parsed.
+func (b *Builder) getDockerMounts() []string {
+	// Returns the longest substring of path that is the path to a directory.
+	var getMaxPath func(path string) string
+	getMaxPath = func(path string) string {
+		f, err := os.Stat(path)
+		if os.IsNotExist(err) {
+			// Try again on parent. This happens because some values in config
+			// are paths to files that get created by the commands, but their
+			// parent directory exists and needs to be mounted.
+			path = filepath.Dir(path)
+			return getMaxPath(path)
+		} else if err != nil {
+			return ""
+		}
+		if f.Mode().IsDir() {
+			return path
+		}
+		return filepath.Dir(path)
+	}
+
+	wd, _ := os.Getwd()
+	mounts := []string{wd}
+
+	config := reflect.ValueOf(b.Config.Builder)
+	for i := 0; i < config.NumField(); i++ {
+		field := getMaxPath(config.Field(i).String())
+		if !strings.HasPrefix(field, "/") {
+			continue
+		}
+		mounts = append(mounts, field)
+	}
+	config = reflect.ValueOf(b.Config.Mixer)
+	for i := 0; i < config.NumField(); i++ {
+		field := getMaxPath(config.Field(i).String())
+		if !strings.HasPrefix(field, "/") {
+			continue
+		}
+		mounts = append(mounts, field)
+	}
+
+	return reduceDockerMounts(mounts)
+}
+
+// RunCommandInContainer will pull the content necessary to build a docker
+// image capable of running the desired command, build that image, and then
+// run the command in that image.
+func (b *Builder) RunCommandInContainer(cmd []string) error {
+	format, first, _, err := b.getUpstreamFormatRange(b.UpstreamVer)
+	if err != nil {
+		return err
+	}
+
+	if err := b.buildDockerImage(format, fmt.Sprint(first)); err != nil {
+		return err
+	}
+
+	// Run command
+	fmt.Printf("Running command in container: %v\n", cmd)
+
+	wd, _ := os.Getwd()
+
+	// Build Docker image
+	dockerCmd := []string{
+		"docker",
+		"run",
+		"-i",
+		"--network=host",
+		"--rm",
+		"--workdir", wd,
+		"--entrypoint", cmd[0],
+	}
+
+	mounts := b.getDockerMounts()
+	for _, path := range mounts {
+		dockerCmd = append(dockerCmd, "-v", fmt.Sprintf("%s:%s", path, path))
+	}
+
+	dockerCmd = append(dockerCmd, getDockerImageName(format))
+	dockerCmd = append(dockerCmd, cmd[1:]...)
+	// TODO: Figure out chicken-egg problem of when to activate this flag. Until
```

Comment:
Update this comment once we get an answer from DevOps about what images we can have built - always latest/out of cycle latest.